src/compiler/crystal/zero_one_or_many.cr

```diff
@@ -0,0 +1,110 @@
+# An Array(T)-like type that's optimized for the case of
+# frequently having zero or one elements.
+#
+# It's a struct, so if you have a ZeroOneOrMany and you
+# want to mutate, you have to use `with` or `without`
+# and overwrite the value you have with what you get
+# from those methods.
+#
+# Warning: you should never keep the original value
+# around when using the above methods as the internal
+# array might be mutated (for performance reasons),
+# so the original value might mutate in that case!
+#
+# TODO: add tests for this type.
+struct Crystal::ZeroOneOrMany(T)
+  include Enumerable(T)
+
+  getter value : Nil | T | Array(T)
+
+  def initialize
+    @value = nil
+  end
+
+  def initialize(@value : T)
+  end
+
+  def initialize(values : Array(T))
+    @value =
+      case values.size
+      when 0
+        nil
+      when 1
+        values.first
+      else
+        values
+      end
+  end
+
+  def each(& : T ->)
+    value = @value
+    case value
+    in Nil
+      # Nothing to do
+    in T
+      yield value
+    in Array(T)
+      value.each do |element|
+        yield element
+      end
+    end
+  end
+
+  def size : Int32
+    value = @value
+    case value
+    in Nil
+      0
+    in T
+      1
+    in Array(T)
+      value.size
+    end
+  end
+
+  def with(element : T) : ZeroOneOrMany(T)
+    value = @value
+    case value
+    in Nil
+      ZeroOneOrMany(T).new(element)
+    in T
+      ZeroOneOrMany(T).new([value, element] of T)
+    in Array(T)
+      value.push element
+      self
+    end
+  end
+
+  def with(elements : Enumerable(T)) : ZeroOneOrMany(T)
+    value = @value
+    case value
+    in Nil
+      ZeroOneOrMany(T).new(elements.map(&.as(T)))
+    in T
+      new_elements = Array(T).new(elements.size + 1)
+      new_elements.push(value)
+      new_elements.concat(elements)
+      ZeroOneOrMany.new(new_elements)
+    in Array(T)
+      value.concat(elements)
+      self
+    end
+  end
+
+  def reject(&block : T -> _)
+    value = @value
+    case value
+    in Nil
+      self
+    in T
+      if yield value
+        ZeroOneOrMany(T).new
+      else
+        self
+      end
+    in Array(T)
+      value.reject! { |element| yield element }
```

Comment:
This `values.reject!` call is dangerous. It affects the value of the original instance, whereas `Enumerable#reject` is expected to return a new collection without altering the existing one.
Example: 
```cr
a = Crystal::ZeroOneOrMany.new(["foo", "bar", "baz"])
b = a.reject &.starts_with?("f")
b # => Crystal::ZeroOneOrMany(String)(@value=["bar", "baz"])
a # => Crystal::ZeroOneOrMany(String)(@value=["bar", "baz"])
```

Altering the original value might often not be an issue, so the behaviour is probably good - it avoids an unnecessary allocation. But it might be better to call this method `#reject!` to make this side effect apparent?

Reply:
> It affects the value of the original instance

Right, I documented this in the type's doc:

```
# Warning: you should never keep the original value
# around when using the above methods as the internal
# array might be mutated (for performance reasons),
# so the original value might mutate in that case!
```

Reply:
If you use `with`, you should always do it like this:

```
some_zero_one_or_many = some_zero_one_or_many.with(...)
```

Alternatively, I'm thinking that instead of `with` we could maybe make it a method that mutates self. Then you have to be careful to always reassign the original value. Maybe it's better?

Reply:
Ah right. I understood "the original value" to refer to an array argument passed to `.new`. But now I see how it connects to the "above methods" as `with` (now `+`) and `without` (which isn't implemented...?).

It probably makes sense to keep it that way.
If there's a use case for keeping an unaltered copy of the original, we can override `#dup` for that.